pkgs/build-support/singularity-tools/default.nix

```diff
@@ -109,14 +82,280 @@ rec {
+      '';
+    in
+    runCommand "${projectName}-image-${name}.sif"
+      {
+        buildInputs = [ singularity util-linux gawk ];
+        passthru = {
+          sandbox = if (sandbox != "") then sandbox else sandboxFromContents;
+          layerClosure = if (sandbox != "") then sandbox.layerClosure else null;
+          inherit singularity;
+        };
+      }
+      ''
+        runHook preImageBuild
+        "${buildscriptPackage}/bin/${buildscriptPackage.meta.mainProgram}" "$out"
+        runHook postImageBuild
+      '';
+
+  buildImageInLinuxVM =
+    { diskSize ? 1024
+    , memSize ? 512
+    }:
+    image:
+    let
+      projectName = image.projectName or image.singularity.projectName or "singularity";
+    in
+    vmTools.runInLinuxVM (image.overrideAttrs (prevAttrs: {
+      buildInputs = prevAttrs.buildInputs or [ ] ++ [
+        e2fsprogs
+      ];
+      preVM = vmTools.createEmptyImage {
+        size = diskSize;
+        fullName = "${projectName}-run-disk";
+      };
+      preImageBuild = ''
+        rm -rf $out
+        mkdir disk
+        mkfs -t ext3 -b 4096 /dev/${vmTools.hd}
+        mount /dev/${vmTools.hd} disk
+        mkdir -p /var/lib/${projectName}/mnt/{container,final,overlay,session,source}
+        echo "root:x:0:0:System administrator:/root:/bin/sh" > /etc/passwd
+        echo > /etc/resolv.conf
+        TMPDIR="$(realpath disk)"; export TMPDIR
+        cd disk
+      '';
+    }));
+
+  buildImage =
+    { name
+    , contents ? [ ]
+    , diskSize ? 1024
+    , runScript ? "#!${stdenv.shell}\nexec /bin/sh"
+    , runAsRoot ? null
+    , memSize ? 512
+    , singularity ? defaultSingularity
+    }:
+    buildImageInLinuxVM
+      {
+        inherit diskSize memSize;
+      }
+      (buildImageFromSandbox {
+        inherit name contents runScript runAsRoot singularity;
+      });
+
+  # shellScript = name: text:
+  #   writeScript name ''
+  #     #!${runtimeShell}
+  #     set -e
+  #     ${text}
+  #   '';
 
-            cd ..
-            mkdir -p /var/lib/${projectName}/mnt/{container,final,overlay,session,source}
-            echo "root:x:0:0:System administrator:/root:/bin/sh" > /etc/passwd
-            echo > /etc/resolv.conf
-            TMPDIR=$(pwd -P) ${projectName} build $out ./img
-          '');
+  # buildImage =
+  #   let
+  #     defaultSingularity = singularity;
+  #   in
+  #   { name
+  #   , contents ? [ ]
+  #   , diskSize ? 1024
+  #   , runScript ? "#!${stdenv.shell}\nexec /bin/sh"
+  #   , runAsRoot ? null
+  #   , memSize ? 512
+  #   , singularity ? defaultSingularity
+  #   }:
+  #   let
+  #     projectName = singularity.projectName or "singularity";
+  #     runAsRootFile = shellScript "run-as-root.sh" runAsRoot;
+  #     runScriptFile = shellScript "run-script.sh" runScript;
+  #     result = vmTools.runInLinuxVM (
+  #       runCommand "${projectName}-image-${name}.sif"
+  #         {
+  #           buildInputs = [ singularity e2fsprogs util-linux gawk ];
```

Comment:
What's the `gawk` in `buildInputs` for?

Reply:
It was used during disk image size calculations, but that was removed in a refactor and the dependency was overlooked.

Reply:
I see. I'll remove it from `buildImageInLinuxVM`.